File: demos/01-essentials/01-ai-foundry/05-rag/rag-azure-py/text_chunker.py
import re
from typing import Dict, List


class TextChunker:
    def __init__(self, chunk_size: int = 1000, chunk_overlap: int = 200):
        self.chunk_size = chunk_size
        self.chunk_overlap = chunk_overlap
    
    def clean_text(self, text: str) -> str:
        text = re.sub(r'\s+', ' ', text)
        text = re.sub(r'\n+', '\n', text)
        return text.strip()
    
    def chunk_text_for_search(self, text: str, metadata: Dict) -> List[Dict]:
        text = self.clean_text(text)
        chunks = []
        
        if len(text) <= self.chunk_size:
            return [{
                'content': text,
                'chunk_id': 0,
                'chunk_count': 1,
                'metadata': metadata.copy()
            }]
        
        start = 0
        chunk_id = 0
        
        while start < len(text):
            end = start + self.chunk_size
            
            if end < len(text):
                sentence_end = text.rfind('.', start, end)
                if sentence_end > start:
                    end = sentence_end + 1
            
            chunk_text = text[start:end].strip()
            
            if chunk_text:
                chunks.append({
                    'content': chunk_text,
                    'chunk_id': chunk_id,
                    'chunk_count': 0,
                    'metadata': metadata.copy()
                })
                chunk_id += 1
            
            start = max(start + self.chunk_size - self.chunk_overlap, end)
        
        for chunk in chunks:
            chunk['chunk_count'] = len(chunks)
        
        return chunks

File: demos/01-essentials/01-ai-foundry/05-rag/rag-azure-cs/TextChunker.cs
using System.Text.RegularExpressions;

namespace RagAzure;

public class TextChunker
{
    private readonly int _chunkSize;
    private readonly int _chunkOverlap;

    public TextChunker(int chunkSize = 1000, int chunkOverlap = 200)
    {
        _chunkSize = chunkSize;
        _chunkOverlap = chunkOverlap;
    }

    public string CleanText(string text)
    {
        text = Regex.Replace(text, @"\s+", " ");
        text = Regex.Replace(text, @"\n+", "\n");
        return text.Trim();
    }

    public List<ChunkResult> ChunkTextForSearch(string text, Dictionary<string, string> metadata)
    {
        text = CleanText(text);
        var chunks = new List<ChunkResult>();

        if (text.Length <= _chunkSize)
        {
            return new List<ChunkResult>
            {
                new ChunkResult
                {
                    Content = text,
                    ChunkId = 0,
                    ChunkCount = 1,
                    Metadata = new Dictionary<string, string>(metadata)
                }
            };
        }

        int start = 0;
        int chunkId = 0;

        while (start < text.Length)
        {
            int end = Math.Min(start + _chunkSize, text.Length);

            if (end < text.Length)
            {
                int sentenceEnd = text.LastIndexOf('.', end - 1, end - start);
                if (sentenceEnd > start)
                {
                    end = sentenceEnd + 1;
                }
            }

            string chunkText = text.Substring(start, end - start).Trim();

            if (!string.IsNullOrWhiteSpace(chunkText))
            {
                chunks.Add(new ChunkResult
                {
                    Content = chunkText,
                    ChunkId = chunkId,
                    ChunkCount = 0, // Will be updated after all chunks are created
                    Metadata = new Dictionary<string, string>(metadata)
                });
                chunkId++;
            }

            start = Math.Max(start + _chunkSize - _chunkOverlap, end);
        }

        // Update chunk count for all chunks
        foreach (var chunk in chunks)
        {
            chunk.ChunkCount = chunks.Count;
        }

        return chunks;
    }
}

public class ChunkResult
{
    public string Content { get; set; } = string.Empty;
    public int ChunkId { get; set; }
    public int ChunkCount { get; set; }
    public Dictionary<string, string> Metadata { get; set; } = new();
}
